File: server/app/core/models/compliance.py
from enum import Enum
from typing import Optional, List, Any
from pydantic import BaseModel, field_validator
from uuid import UUID
from datetime import datetime, date


class ComplianceCategory(str, Enum):
    # Labor
    minimum_wage = "minimum_wage"
    overtime = "overtime"
    sick_leave = "sick_leave"
    meal_breaks = "meal_breaks"
    pay_frequency = "pay_frequency"
    final_pay = "final_pay"
    minor_work_permit = "minor_work_permit"
    scheduling_reporting = "scheduling_reporting"
    workers_comp = "workers_comp"
    business_license = "business_license"
    tax_rate = "tax_rate"
    posting_requirements = "posting_requirements"
    leave = "leave"
    workplace_safety = "workplace_safety"
    anti_discrimination = "anti_discrimination"
    # Healthcare
    hipaa_privacy = "hipaa_privacy"
    billing_integrity = "billing_integrity"
    clinical_safety = "clinical_safety"
    healthcare_workforce = "healthcare_workforce"
    corporate_integrity = "corporate_integrity"
    research_consent = "research_consent"
    state_licensing = "state_licensing"
    emergency_preparedness = "emergency_preparedness"
    # Oncology
    radiation_safety = "radiation_safety"
    chemotherapy_handling = "chemotherapy_handling"
    tumor_registry = "tumor_registry"
    oncology_clinical_trials = "oncology_clinical_trials"
    oncology_patient_rights = "oncology_patient_rights"
    # Medical compliance
    health_it = "health_it"
    quality_reporting = "quality_reporting"
    cybersecurity = "cybersecurity"
    environmental_safety = "environmental_safety"
    pharmacy_drugs = "pharmacy_drugs"
    payer_relations = "payer_relations"
    reproductive_behavioral = "reproductive_behavioral"
    pediatric_vulnerable = "pediatric_vulnerable"
    telehealth = "telehealth"
    medical_devices = "medical_devices"
    transplant_organ = "transplant_organ"
    antitrust = "antitrust"
    tax_exempt = "tax_exempt"
    language_access = "language_access"
    records_retention = "records_retention"
    marketing_comms = "marketing_comms"
    emerging_regulatory = "emerging_regulatory"
    # Life Sciences
    gmp_manufacturing = "gmp_manufacturing"
    glp_nonclinical = "glp_nonclinical"
    clinical_trials_gcp = "clinical_trials_gcp"
    drug_supply_chain = "drug_supply_chain"
    sunshine_open_payments = "sunshine_open_payments"
    biosafety_lab = "biosafety_lab"


class JurisdictionLevel(str, Enum):
    federal = "federal"
    state = "state"
    county = "county"
    city = "city"
    special_district = "special_district"
    regulatory_body = "regulatory_body"
    # International levels
    national = "national"       # Country-level law
    province = "province"       # Subnational division (international)
    region = "region"           # UK constituent countries, etc.


class AlertSeverity(str, Enum):
    info = "info"
    warning = "warning"
    critical = "critical"


class AlertStatus(str, Enum):
    unread = "unread"
    read = "read"
    dismissed = "dismissed"
    actioned = "actioned"


class AlertType(str, Enum):
    change = "change"
    new_requirement = "new_requirement"
    upcoming_legislation = "upcoming_legislation"
    deadline_approaching = "deadline_approaching"


class LegislationStatus(str, Enum):
    proposed = "proposed"
    passed = "passed"
    signed = "signed"
    effective_soon = "effective_soon"
    effective = "effective"
    dismissed = "dismissed"


class BusinessLocation(BaseModel):
    id: UUID
    company_id: UUID
    jurisdiction_id: Optional[UUID] = None
    name: Optional[str] = None
    address: Optional[str] = None
    city: str
    state: Optional[str] = None
    county: Optional[str] = None
    zipcode: Optional[str] = None
    country_code: str = "US"
    is_active: bool = True
    auto_check_enabled: bool = True
    auto_check_interval_days: int = 7
    next_auto_check: Optional[datetime] = None
    last_compliance_check: Optional[datetime] = None
    has_local_ordinance: Optional[bool] = None
    source: Optional[str] = "manual"
    coverage_status: Optional[str] = "covered"
    employee_count: Optional[int] = 0
    facility_attributes: Optional[dict] = None
    created_at: datetime
    updated_at: datetime

    @field_validator("facility_attributes", mode="before")
    @classmethod
    def parse_facility_attributes(cls, v):
        if isinstance(v, str):
            import json
            return json.loads(v)
        return v


class ComplianceRequirement(BaseModel):
    id: UUID
    location_id: UUID
    category: str
    jurisdiction_level: str
    jurisdiction_name: str
    title: str
    description: Optional[str] = None
    current_value: Optional[str] = None
    numeric_value: Optional[float] = None
    source_url: Optional[str] = None
    source_name: Optional[str] = None
    effective_date: Optional[date] = None
    expiration_date: Optional[date] = None
    previous_value: Optional[str] = None
    last_changed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class ComplianceAlert(BaseModel):
    id: UUID
    location_id: UUID
    company_id: UUID
    requirement_id: Optional[UUID] = None
    title: str
    message: str
    severity: str
    status: str
    category: Optional[str] = None
    action_required: Optional[str] = None
    source_url: Optional[str] = None
    source_name: Optional[str] = None
    deadline: Optional[date] = None
    confidence_score: Optional[float] = None
    verification_sources: Optional[list] = None
    alert_type: Optional[str] = "change"
    effective_date: Optional[date] = None
    metadata: Optional[dict] = None
    created_at: datetime
    read_at: Optional[datetime] = None
    dismissed_at: Optional[datetime] = None


class LocationCreate(BaseModel):
    name: Optional[str] = None
    address: Optional[str] = None
    city: str
    state: Optional[str] = None
    county: Optional[str] = None
    zipcode: Optional[str] = None
    country_code: str = "US"
    facility_attributes: Optional[dict] = None


class LocationUpdate(BaseModel):
    name: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    county: Optional[str] = None
    zipcode: Optional[str] = None
    is_active: Optional[bool] = None


class FacilityAttributesUpdate(BaseModel):
    entity_type: Optional[str] = None
    payer_contracts: Optional[List[str]] = None
    bed_count: Optional[int] = None
    teaching_hospital: Optional[bool] = None


class AutoCheckSettings(BaseModel):
    auto_check_enabled: Optional[bool] = None
    auto_check_interval_days: Optional[int] = None


class RequirementResponse(BaseModel):
    id: str
    category: str
    rate_type: Optional[str] = None
    applicable_industries: Optional[List[str]] = None
    jurisdiction_level: str
    jurisdiction_name: str
    title: str
    description: Optional[str] = None
    current_value: Optional[str] = None
    numeric_value: Optional[float] = None
    source_url: Optional[str] = None
    source_name: Optional[str] = None
    effective_date: Optional[str] = None
    previous_value: Optional[str] = None
    last_changed_at: Optional[str] = None
    affected_employee_count: Optional[int] = None
    affected_employee_names: Optional[List[str]] = None
    min_wage_violation_count: Optional[int] = None
    is_pinned: bool = False


class PinRequirementRequest(BaseModel):
    is_pinned: bool = True


class AlertResponse(BaseModel):
    id: str
    location_id: str
    requirement_id: Optional[str] = None
    title: str
    message: str
    severity: str
    status: str
    category: Optional[str] = None
    action_required: Optional[str] = None
    source_url: Optional[str] = None
    source_name: Optional[str] = None
    deadline: Optional[str] = None
    confidence_score: Optional[float] = None
    verification_sources: Optional[list] = None
    alert_type: Optional[str] = None
    effective_date: Optional[str] = None
    metadata: Optional[dict] = None
    impact_summary: Optional[str] = None
    affected_employee_count: Optional[int] = None
    created_at: str
    read_at: Optional[str] = None


class CalendarItem(BaseModel):
    """Compliance calendar row — a non-dismissed alert with a deadline,
    enriched with location context and a status bucket the UI can group by.
    """
    id: str
    location_id: str
    location_name: Optional[str] = None
    location_state: Optional[str] = None
    jurisdiction_name: Optional[str] = None
    requirement_id: Optional[str] = None
    title: str
    category: Optional[str] = None
    severity: str
    deadline: str  # ISO date — calendar items always have one
    derived_status: str  # 'overdue' | 'due_soon' | 'upcoming' | 'future'
    days_until_due: int
    action_required: Optional[str] = None
    alert_status: str  # underlying alert status (unread / read / actioned)
    created_at: str


class PayerPolicyResponse(BaseModel):
    id: str
    payer_name: str
    payer_type: Optional[str] = None
    policy_number: Optional[str] = None
    policy_title: Optional[str] = None
    procedure_codes: list[str] = []
    procedure_description: Optional[str] = None
    coverage_status: str = "conditional"
    requires_prior_auth: bool = False
    clinical_criteria: Optional[str] = None
    documentation_requirements: Optional[str] = None
    medical_necessity_criteria: Optional[str] = None
    age_restrictions: Optional[str] = None
    frequency_limits: Optional[str] = None
    source_url: Optional[str] = None
    source_document: Optional[str] = None
    effective_date: Optional[str] = None
    last_reviewed: Optional[str] = None


class CheckLogEntry(BaseModel):
    id: str
    location_id: str
    company_id: str
    check_type: str
    status: str
    started_at: str
    completed_at: Optional[str] = None
    new_count: int = 0
    updated_count: int = 0
    alert_count: int = 0
    error_message: Optional[str] = None


class UpcomingLegislationResponse(BaseModel):
    id: str
    location_id: str
    category: Optional[str] = None
    title: str
    description: Optional[str] = None
    current_status: str
    expected_effective_date: Optional[str] = None
    impact_summary: Optional[str] = None
    source_url: Optional[str] = None
    source_name: Optional[str] = None
    confidence: Optional[float] = None
    days_until_effective: Optional[int] = None
    affected_employee_count: Optional[int] = None
    created_at: str


class VerificationResult(BaseModel):
    confirmed: bool
    confidence: float
    sources: List[dict] = []
    explanation: str = ""


class ComplianceSummary(BaseModel):
    total_locations: int
    total_requirements: int
    unread_alerts: int
    critical_alerts: int
    recent_changes: list
    auto_check_locations: int = 0
    upcoming_deadlines: list = []


# ── Hierarchical Compliance Response Schemas ──────────────────────────────
# ALL intelligence is computed server-side. Frontend just renders these.


class TriggerActivation(BaseModel):
    """Server-evaluated trigger result — frontend just displays this."""
    trigger_type: str
    trigger_key: Optional[str] = None
    trigger_value: Optional[Any] = None
    matched: bool = True


class JurisdictionLevelRequirement(BaseModel):
    """A single requirement at one jurisdiction level. Server has already
    determined status, evaluated triggers, and resolved citations."""
    id: str
    jurisdiction_level: str
    jurisdiction_name: str
    title: str
    description: Optional[str] = None
    current_value: Optional[str] = None
    numeric_value: Optional[float] = None
    source_url: Optional[str] = None
    statute_citation: Optional[str] = None
    status: str = "active"
    canonical_key: Optional[str] = None
    triggered_by: Optional[List[TriggerActivation]] = None


class PrecedenceInfo(BaseModel):
    """Server-resolved precedence determination."""
    precedence_type: str
    reasoning_text: Optional[str] = None
    legal_citation: Optional[str] = None
    trigger_condition: Optional[dict] = None


class CategoryComplianceStack(BaseModel):
    """Fully resolved compliance stack for one category."""
    category: str
    category_label: str
    domain: Optional[str] = None
    authority_type: Optional[str] = None
    governing_level: str
    governing_requirement: JurisdictionLevelRequirement
    precedence: Optional[PrecedenceInfo] = None
    all_levels: List[JurisdictionLevelRequirement]
    affected_employee_count: Optional[int] = None


class HierarchicalComplianceResponse(BaseModel):
    """Complete server-resolved compliance view for a location.
    Every field is final — no client-side post-processing needed."""
    location_id: str
    location_name: str
    city: str
    state: str
    facility_attributes: Optional[dict] = None
    categories: List[CategoryComplianceStack]
    total_categories: int
    total_requirements: int

File: client/src/types/compliance.ts
import { CATEGORY_LABELS } from '../generated/complianceCategories'
export { CATEGORY_LABELS }
export type { CategoryGroup } from '../generated/complianceCategories'

export interface JurisdictionOption {
  city: string
  state: string
  county: string | null
  has_local_ordinance: boolean
}

/** @deprecated Use JurisdictionOption */
export type Jurisdiction = JurisdictionOption

export interface FacilityAttributes {
  entity_type?: string | null
  payer_contracts?: string[] | null
  bed_count?: number | null
  teaching_hospital?: boolean | null
}

/**
 * Clinical-care entity types. Presence of one on any active location is what
 * makes the payer/protocol/policy compliance tabs relevant. A dental office,
 * pharmacy, or lab is `entity_type` too but does NOT get those tabs — the gate
 * is this explicit allowlist, not "has an entity_type" or "has payer_contracts"
 * (a dental office can carry payer contracts and still not need payer policies).
 */
export const CLINICAL_ENTITY_TYPES: ReadonlySet<string> = new Set([
  'fqhc',
  'hospital',
  'critical_access_hospital',
  'clinic',
  'nursing_facility',
  'behavioral_health',
  'ambulatory_surgery_center',
  'home_health',
  'hospice',
  'dialysis_center',
])

// ── Risk cockpit ──────────────────────────────────────────────────────────
export interface RiskPenalty {
  civil_min?: number | null
  civil_max?: number | null
  per_violation?: boolean | null
  annual_cap?: number | null
  enforcing_agency?: string | null
  summary?: string | null
}

export interface RiskIssue {
  id: string
  source: 'wage' | 'credential' | 'incident' | 'alert'
  severity: 'critical' | 'high' | 'moderate'
  title: string
  detail?: string | null
  employee_names: string[]
  location_label?: string | null
  penalty?: RiskPenalty | null
  statute_citation?: string | null
  recommendation?: string | null
  link?: string | null
  deadline?: string | null
  alert_id?: string | null
}

export interface RiskPosture {
  open_critical: number
  open_high: number
  open_moderate: number
  employees_affected: number
  exposure_min_usd: number
  exposure_max_usd: number
  exposure_unquantified_count: number
  next_deadline_days?: number | null
  next_deadline_label?: string | null
}

export interface RiskGetAhead {
  title: string
  kind: 'legislation' | 'deadline'
  effective_date?: string | null
  days_until?: number | null
  location_label?: string | null
}

export interface ComplianceRiskSummary {
  posture: RiskPosture
  issues: RiskIssue[]
  get_ahead: RiskGetAhead[]
  generated_at: string
}

export interface LocationCreate {
  name?: string
  address?: string
  city: string
  state: string
  county?: string
  zipcode?: string
  facility_attributes?: FacilityAttributes
  ein?: string
  naics?: string
  max_employees?: number
  annual_avg_employees?: number
}

export interface LocationUpdate {
  name?: string
  address?: string
  city?: string
  state?: string
  county?: string
  zipcode?: string
  is_active?: boolean
  ein?: string
  naics?: string
  max_employees?: number
  annual_avg_employees?: number
}

export interface BusinessLocation {
  id: string
  company_id: string
  name: string | null
  address: string | null
  city: string
  state: string
  county: string | null
  zipcode: string | null
  is_active: boolean
  auto_check_enabled: boolean
  auto_check_interval_days: number
  next_auto_check: string | null
  last_compliance_check: string | null
  requirements_count: number
  unread_alerts_count: number
  employee_count: number
  employee_names: string[]
  data_status: string
  coverage_status: string
  has_local_ordinance: boolean
  facility_attributes?: FacilityAttributes | null
  source?: 'manual' | 'employee_derived'
  ein: string | null
  naics: string | null
  max_employees: number | null
  annual_avg_employees: number | null
  created_at: string
}

export interface ComplianceRequirement {
  id: string
  category: string
  rate_type: string | null
  applicable_industries: string[]
  jurisdiction_level: 'federal' | 'state' | 'county' | 'city' | 'special_district' | 'regulatory_body'
  jurisdiction_name: string
  title: string
  description: string | null
  current_value: string | null
  numeric_value: number | null
  source_url: string | null
  /** Liveness of source_url: 'unchecked' | 'ok' | 'dead' (null = no catalog link). */
  source_url_status?: 'unchecked' | 'ok' | 'dead' | null
  /** The OPERATIVE statute for this row's value, verified against the authority's own text (null = not yet codified). */
  statute_citation?: string | null
  citation_verified_at?: string | null
  /** Higher-level authorities this row sits on top of rather than restates — e.g. the federal floor a state threshold must meet or exceed. */
  jurisdictional_basis?: Array<{
    citation: string
    item_id: string
    index_slug: string | null
    level: string
    relation: 'floor'
  }> | null
  source_name: string | null
  effective_date: string | null
  previous_value: string | null
  last_changed_at: string | null
  affected_employee_count: number | null
  affected_employee_names: string[]
  min_wage_violation_count: number | null
  is_pinned: boolean
}

export interface VerificationSource {
  url: string
  name: string
  type: 'official' | 'news' | 'blog' | 'other'
  snippet?: string
}

export interface ComplianceAlert {
  id: string
  location_id: string
  requirement_id: string | null
  alert_type: 'change' | 'new_requirement' | 'upcoming_legislation' | 'deadline_approaching' | null
  title: string
  message: string
  severity: 'info' | 'warning' | 'critical'
  status: 'unread' | 'read' | 'dismissed' | 'actioned'
  category: string | null
  action_required: string | null
  source_url: string | null
  source_name: string | null
  deadline: string | null
  confidence_score: number | null
  verification_sources: VerificationSource[] | null
  effective_date: string | null
  metadata: Record<string, unknown> | null
  impact_summary: string | null
  affected_employee_count: number | null
  created_at: string
  read_at: string | null
}

export interface ComplianceSummary {
  total_locations: number
  total_requirements: number
  unread_alerts: number
  critical_alerts: number
  recent_changes: {
    location: string
    category: string
    title: string
    old_value: string | null
    new_value: string
    changed_at: string
  }[]
  auto_check_locations: number
  upcoming_deadlines: {
    title: string
    effective_date: string
    days_until: number
    status: string
    category: string | null
    location: string
  }[]
}

export interface PinnedRequirement {
  id: string
  category: string
  jurisdiction_level: string
  jurisdiction_name: string
  title: string
  description: string | null
  current_value: string | null
  effective_date: string | null
  source_url: string | null
  is_pinned: boolean
  location_name: string | null
  city: string
  state: string
}

export interface CheckLogEntry {
  id: string
  location_id: string
  company_id: string
  check_type: 'manual' | 'scheduled' | 'proactive'
  status: 'running' | 'completed' | 'failed'
  started_at: string
  completed_at: string | null
  new_count: number
  updated_count: number
  alert_count: number
  error_message: string | null
}

export interface UpcomingLegislation {
  id: string
  location_id: string
  category: string | null
  title: string
  description: string | null
  current_status: 'proposed' | 'passed' | 'signed' | 'effective_soon' | 'effective' | 'dismissed'
  expected_effective_date: string | null
  impact_summary: string | null
  source_url: string | null
  source_name: string | null
  confidence: number | null
  days_until_effective: number | null
  affected_employee_count: number | null
  created_at: string
}

export interface AssignableUser {
  id: string
  name: string
  email: string
  role: 'client' | 'admin'
}

export interface ComplianceActionPlanUpdate {
  action_owner_id?: string | null
  next_action?: string | null
  action_due_date?: string | null
  recommended_playbook?: string | null
  estimated_financial_impact?: string | null
  mark_actioned?: boolean
}

export interface AvailablePoster {
  location_id: string
  location_name: string
  state: string
  template_id: string | null
  poster_type: string
  title: string
  description: string | null
  download_url: string | null
  status: 'available' | 'ordered' | 'not_available'
}

export interface PosterOrder {
  id: string
  location_id: string
  template_ids: string[]
  status: 'pending' | 'processing' | 'shipped' | 'delivered' | 'cancelled'
  created_at: string
}

// SSE event types for compliance check streaming

export interface SSEHeartbeat {
  type: 'heartbeat'
}

export interface SSEProgress {
  type: 'progress'
  message: string
  phase: string
  progress: number
  timestamp: string
}

export interface SSENewRequirement {
  type: 'new_requirement'
  requirement_id: string
  category: string
  title: string
  jurisdiction_level: string
  current_value: string
}

export interface SSEAlertGenerated {
  type: 'alert_generated'
  alert_id: string
  title: string
  severity: string
  message: string
}

export interface SSEError {
  type: 'error'
  message: string
}

export type ComplianceSSEEvent =
  | SSEHeartbeat
  | SSEProgress
  | SSENewRequirement
  | SSEAlertGenerated
  | SSEError

// Requirement categories

export const REQUIREMENT_CATEGORIES = [
  'minimum_wage',
  'overtime',
  'sick_leave',
  'family_leave',
  'anti_discrimination',
  'workplace_safety',
  'workers_comp',
  'tax_withholding',
  'pay_frequency',
  'meal_breaks',
  'rest_breaks',
  'final_pay',
  'posting_requirements',
  'pto',
  'fair_scheduling',
  'ban_the_box',
  'non_compete',
  'harassment_training',
  'data_privacy',
  'whistleblower',
  'accommodations',
  'other',
] as const

export type RequirementCategory = (typeof REQUIREMENT_CATEGORIES)[number]

// ── Hierarchical Compliance Response Types ───────────────────────────────
// ALL intelligence is computed server-side. These are display-only types.

export interface TriggerActivation {
  trigger_type: 'attribute' | 'entity_type' | 'requirement_active' | 'category_active' | 'none'
  trigger_key: string | null
  trigger_value: unknown | null
  matched: boolean
}

export interface JurisdictionLevelRequirement {
  id: string
  jurisdiction_level: string
  jurisdiction_name: string
  title: string
  description: string | null
  current_value: string | null
  numeric_value: number | null
  source_url: string | null
  /** Liveness of source_url: 'unchecked' | 'ok' | 'dead'. */
  source_url_status?: 'unchecked' | 'ok' | 'dead' | null
  statute_citation: string | null
  status: string
  canonical_key: string | null
  triggered_by: TriggerActivation[] | null
}

export interface PrecedenceInfo {
  precedence_type: 'floor' | 'ceiling' | 'supersede' | 'additive'
  reasoning_text: string | null
  legal_citation: string | null
  trigger_condition: Record<string, unknown> | null
}

export interface CategoryComplianceStack {
  category: string
  category_label: string
  domain: string | null
  authority_type: string | null
  governing_level: string
  governing_requirement: JurisdictionLevelRequirement
  precedence: PrecedenceInfo | null
  all_levels: JurisdictionLevelRequirement[]
  affected_employee_count: number | null
}

export interface HierarchicalComplianceResponse {
  location_id: string
  location_name: string
  city: string
  state: string
  facility_attributes: Record<string, unknown> | null
  categories: CategoryComplianceStack[]
  total_categories: number
  total_requirements: number
}

export const categoryLabel: Record<RequirementCategory, string> = {
  minimum_wage: 'Minimum Wage',
  overtime: 'Overtime',
  sick_leave: 'Sick Leave',
  family_leave: 'Family Leave',
  anti_discrimination: 'Anti-Discrimination',
  workplace_safety: 'Workplace Safety',
  workers_comp: "Workers' Comp",
  tax_withholding: 'Tax Withholding',
  pay_frequency: 'Pay Frequency',
  meal_breaks: 'Meal Breaks',
  rest_breaks: 'Rest Breaks',
  final_pay: 'Final Pay',
  posting_requirements: 'Posting Requirements',
  pto: 'PTO',
  fair_scheduling: 'Fair Scheduling',
  ban_the_box: 'Ban the Box',
  non_compete: 'Non-Compete',
  harassment_training: 'Harassment Training',
  data_privacy: 'Data Privacy',
  whistleblower: 'Whistleblower',
  accommodations: 'Accommodations',
  other: 'Other',
}
